### ISSUE TITLE
Exchange.class returns an array as opposed to JSON object

### ISSUE THREAD

**Issue by usmankhanic** [ccxt/ccxt#4058]
In ccxt/js/base/Exchange.js

line 960 
return [price, amount]

This is an array. If downstream systems are trying to resolve this into an Object, they would have to convert to an array and fetch the values via an Index, which is in-efficient and wont help from an Object typing perspective either. 

Suggest fix

return {price, amount}

This returns a JSON object that can be easily converted to a typed object downstream.

**Comment by usmankhanic** [comment]
It’s directly below the line highlighted 

    parseBidAsk (bidask, priceKey = 0, amountKey = 1) {
        let price = parseFloat (bidask[priceKey])
        let amount = parseFloat (bidask[amountKey])
        return [ price, amount ]
    }

See the return [price, amount] 

Should not that be {price, amount} 

?

I’m on the latest ccxt release and I understand the data structures well. 

Thank you

**Comment by kroitor** [comment]
> Should not that be {price, amount} ?

No, for several reasons:
- this is how it's documented in our Manual
- the majority of exchanges will return arrays, not objects, which is still a perfectly legal JSON, therefore conversion from arrays to objects will take more time than it takes to parse the orderbook response now
- if we use objects for bidasks, then each bidask will have to duplicate keynames (`price`, `amount`), which will effectively explode the orderbook structure in terms of bytesize and serialization

Hope this answers your question.

**Comment by usmankhanic** [comment]
Thanks for the reply - but there is a discrepancy.

The L2Update entity has an array of L2Point. the documentation states that the schema for snapshots and updates is meant to be the same. 

Yet, when snapshotting, as per my point above, you are no longer returning an array of L2Point compatible values. Rather you are returning an array of arrays, which is very different. 

No one has said to use key names  - rather, simply replace the [] with {} e.g  [0.0023, 10.3] with {0.0023, 10.3}  - this will keep both the snapshot and the update consistent - per your manual.

Does that make sense?

**Comment by usmankhanic** [comment]
Also - in the fetchOrderBook method, you are indeed returning an object with keys - "price" and "amount". Where is an an L2update, it's price and size -- > again, inconsistent.. right?

**Comment by kroitor** [comment]
@usmankhanic 

> Also - in the fetchOrderBook method, you are indeed returning an object with keys - "price" and "amount". 

No, we're not returning object with keys from fetchOrderBook, unless there's a bug with some particular exchange. But with the base Exchange class, your above statement isn't true. Proof:

```
mbp:ccxt igorkroitor$ node examples/js/cli binance fetchOrderBook YOYOW/BTC
binance.fetchOrderBook (YOYOW/BTC)
{      bids: [ [0.00000515, 78250],
               [0.00000514, 49191],
               [0.00000513, 11754],
               [0.00000512, 33765],
               [0.00000511, 38335],
               [0.0000051, 47858],
               [0.00000509, 20873],
               [0.00000508, 85838],
               [0.00000507, 27950],
               [0.00000506, 104771],
               [0.00000505, 93260],
               [0.00000504, 17644],
               [0.00000503, 23999],
               [0.00000502, 16769],
               [0.00000501, 235753],
               ...
               [0.00000381, 524]     ],
       asks: [ [0.00000518, 476],
               [0.00000519, 2451],
               [0.0000052, 15194],
               [0.00000521, 6486],
               [0.00000522, 14577],
               [0.00000523, 12461],
               [0.00000524, 3869],
               [0.00000525, 22547],
               [0.00000526, 8083],
               ...
               [0.00000619, 387512]  ],
  timestamp:    undefined,
   datetime:    undefined,
      nonce:    36996470                }
mbp:ccxt igorkroitor$
```

> Where as in L2update, it's price and size -- > again, inconsistent.. right?

Not really sure what is an L2 update, but we return the same format from fetchL2OrderBook:

```
mbp:ccxt igorkroitor$ node examples/js/cli binance fetchL2OrderBook YOYOW/BTC
binance.fetchL2OrderBook (YOYOW/BTC)
{      bids: [ [0.00000515, 77611],
               [0.00000514, 49191],
               [0.00000513, 11754],
               [0.00000512, 34389],
               [0.00000511, 38335],
               [0.0000051, 44727],
               [0.00000509, 36091],
               [0.00000508, 94481],
               [0.00000507, 27950],
               [0.00000506, 89553],
               ...
               [0.00000381, 524]     ],
       asks: [ [0.00000518, 476],
               [0.00000519, 2451],
               [0.0000052, 15194],
               [0.00000521, 6486],
               [0.00000522, 14577],
               [0.00000523, 12461],
               [0.00000524, 3869],
               [0.00000525, 22547],
               [0.00000526, 8083],
               ...
               [0.00000619, 387512]  ],
  timestamp:    undefined,
   datetime:    undefined,
      nonce:    36996528                }
```

So, from the above it follows that we return [ price, amount ] everywhere, so I'm not really sure what you are talking about. Without seeing at least some proper debugging info from you according to these guides (https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-submit-an-issue) it will be really hard to help you.
